### PR TITLE
Feature watchdog transitions

### PR DESCRIPTION
**Pull request by byoms23**
Small refactor to support `watchdog` type for rules when validating actions.

### CODE REVIEW

File: lib/matchers.js [hunk 1]
```diff
@@ -19,25 +22,110 @@ const matchAll = (rule) => {
...
-      ((rule.confidence === undefined && actionsList[i].confidence === undefined) || actionsList[i].confidence > rule.confidence)) {
-        return true
+const matchWatchdog = (domain, rule, action, cache) => {
+  if(cache.wdCalled) {
+    return matchWatchdogActions(rule, cache.wdActions, cache)
+  }
+
+  return watchdogSdk.getIntent(domain, action.value)
+    .then((wdResponse) => {
+      const wdClassification = wdResponse.results[0]
```
Comment (yoelfme) [inline lib/matchers.js]: why you just take the first element in the results?

File: lib/index.js [hunk 2]
```diff
@@ -2,156 +2,159 @@ const _ = require('lodash')
...
-                reject(err)
-              }
-              resolve(newInstance)
-            })
-          })
+  addInstance (instance) {
+    return new Promise((resolve, reject) => {
+      client.set(instance.id, instance, this.model.ttl, (err) => {
+        if (err) {
+          reject(err)
```
Comment (baquiax) [inline lib/index.js]: I think this line must have  a `return` statement because if not `resolve` will be executed too.

File: lib/index.js [hunk 3]
```diff
@@ -2,156 +2,159 @@ const _ = require('lodash')
...
-          // Load globalTransitions and sort all transitions
-          this.model = Utils.prepareModel(model)
+  newInstance (id) {
+    return new Promise((resolve, reject) => {
+      const newState = _.cloneDeep(this.model.states.get(INITIAL_STATE))
+      const newInstance = new Instance(id, this.middlewares, newState)
+      console.log(`Creating new instance %j`, newInstance)
+      client.set(id, newInstance, this.model.ttl, (err) => {
+        if (err) {
+          reject(err)
```
Comment (baquiax) [inline lib/index.js]: It must be `return reject(err)`

File: lib/index.js [hunk 4]
```diff
@@ -2,156 +2,159 @@ const _ = require('lodash')
...
-                  } else {
+  getState (instance, data) {
+    const updatedInstance = _.cloneDeep(instance)
+    return new Promise((resolve, reject) => {
+      if (data && data.action) {
+        this.validateTransition(updatedInstance, data.action)
+          .then((transition) => {
+            console.log('Moving to transition: %j', transition)
+            return this.searchNextState(transition.to || transition, transition.to === undefined)
+              .then((nextState) => {
```
[the remark this answers is not in the document]
Comment (baquiax) [reply lib/index.js]: I think the same.

File: lib/index.js [hunk 5]
```diff
@@ -162,98 +165,116 @@ module.exports.new = () => {
...
-            })
-          })
-        }
+  goToDefault (instance) {
+    return new Promise((resolve, reject) => {
+      this.searchNextState('default', false).then((nextState) => {
+        instance.currentState = nextState
+        client.set(instance.id, instance, this.model.ttl, (err) => {
+          if (err) {
+            reject(err)
```
Comment (baquiax) [inline lib/index.js]: Does it must have a `return` statement?

File: lib/index.js [hunk 6]
```diff
@@ -162,98 +165,116 @@ module.exports.new = () => {
...
+    })
+  }
 
-        on (eventName, eventFunction) {
-          this.internalEmitter.on(eventName, eventFunction)
+  saveInstance (instance) {
+    return new Promise((resolve, reject) => {
+      client.set(instance.id, instance, this.model.ttl, (err) => {
+        if (err) {
+          reject(err)
```
Comment (baquiax) [inline lib/index.js]: Does it must have a `return` statement?

**Pull request merged**